crates/wasmtime/src/runtime/store.rs

```diff
@@ -1047,6 +1067,105 @@ impl<T> Store<T> {
+    /// - Will panic if a pending exception has already been set.
+    #[cfg(feature = "gc")]
+    pub fn throw<R>(&mut self, exception: Rooted<ExnRef>) -> Result<R, ThrownException> {
+        self.inner.throw_impl(exception);
+        Err(ThrownException)
+    }
+
+    /// Take the currently pending exception, if any, and return it,
+    /// removing it from the "pending exception" slot.
+    ///
+    /// If there is no pending exception, returns `None`.
+    ///
+    /// Note: the returned exception is a LIFO root (see
+    /// [`crate::Rooted`]), rooted in the current handle scope. Take
+    /// care to ensure that it is re-rooted or otherwise does not
+    /// escape this scope! It is usually best to allow an exception
+    /// object to be rooted in the store's "pending exception" slot
+    /// until the final consumer has taken it, rather than root it and
+    /// pass it up the callstack in some other way.
+    ///
+    /// This method is useful to implement ad-hoc exception plumbing
+    /// in various ways, but for the most idiomatic handling, see
+    /// [`StoreContextMut::catch`].
+    #[cfg(feature = "gc")]
+    pub fn take_pending_exception(&mut self) -> Option<Rooted<ExnRef>> {
+        self.inner.take_pending_exception_rooted()
+    }
+
+    /// Tests whether there is a pending exception.
+    ///
+    /// Ordinarily, a pending exception will be set on a store if and
+    /// only if a host-side callstack is propagating a
+    /// [`crate::ThrownException`] error. The final consumer that
+    /// catches the exception takes it; it may re-place it to re-throw
+    /// (using [`throw`]) if it chooses not to actually handle the
+    /// exception.
+    ///
+    /// This method is useful to tell whether a store is in this
+    /// state, but should not be used as part of the ordinary
+    /// exception-handling flow. For the most idiomatic handling, see
+    /// [`StoreContextMut::catch`].
+    #[cfg(feature = "gc")]
+    pub fn has_pending_exception(&self) -> bool {
+        self.inner.pending_exception.is_some()
+    }
+
+    /// Invoke a closure, handling any thrown Wasm exception with the given handler.
+    #[cfg(feature = "gc")]
+    pub fn catch<F, Handler, R>(&mut self, f: F, handler: Handler) -> Result<R>
+    where
+        F: FnOnce(StoreContextMut<'_, T>) -> Result<R>,
+        Handler: FnOnce(StoreContextMut<'_, T>, Rooted<ExnRef>) -> Result<R>,
+    {
+        match f(self.as_context_mut()) {
+            Err(e) if e.is::<ThrownException>() => {
+                debug_assert!(self.has_pending_exception());
+                let mut scope = crate::RootScope::new(self);
+                let exn = scope.as_context_mut().take_pending_exception().unwrap();
+                handler(scope.as_context_mut(), exn)
+            }
+            ret @ Ok(_) | ret @ Err(_) => {
+                debug_assert!(!self.has_pending_exception());
+                ret
+            }
+        }
+    }
```

Comment:
Given that we need to in theory replicate these methods on `StoreContextMut`, `Store`, `Caller`, and other things that implement `AsContextMut`, and given that in theory we also need to eventually add `catch_async` in addition to `catch`, I'd subjectively say this is probably fine to relegate to examples or documentation but otherwise not live here as a first-class function.

Reply:
Sure, removed for now -- always possible to add later if it turns out to be a real ergonomic need...